### PR TITLE
Group plugin improvements

### PR DESCRIPTION
**Pull request by jdalsem**
also added group join/leave to the group entity menu

### CODE REVIEW

File: mod/groups/views/default/groups/profile/fields.php [hunk 1]
```diff
@@ -3,32 +3,43 @@
...
+	$options = ['value' => $group->$key];
+	if ($valtype == 'tags') {
+		$options['tag_names'] = $key;
 	}
+	
+	$field_title = elgg_echo("groups:{$key}");
+	$field_value = elgg_view("output/$valtype", $options);
+	
+	echo <<<___FIELD
+	<div class='clearfix group-profile-field'>
```
Comment (hypeJunction) [inline mod/groups/views/default/groups/profile/fields.php]: Just add the view for this.

Comment (hypeJunction) [reply mod/groups/views/default/groups/profile/fields.php]: Don't use clearfix. There are 3 types of containers for grids that control the responsiveness

File: mod/groups/start.php [hunk 2]
```diff
@@ -357,6 +355,82 @@ function groups_activity_owner_block_menu($hook, $type, $return, $params) {
...
+ * @param \ElggUser  $user  User to check leave action for
+ *
+ * @return \ElggMenuItem|false
+ */
+function groups_get_group_leave_menu_item(\ElggGroup $group, $user = null) {
+	if (!$group instanceof \ElggGroup) {
+		return false;
+	}
+	
+	if (is_null($user)) {
```
Comment (jeabakker) [inline mod/groups/start.php]: better is `!$user instanceof ElggUser`

File: mod/groups/start.php [hunk 3]
```diff
@@ -357,6 +355,82 @@ function groups_activity_owner_block_menu($hook, $type, $return, $params) {
...
+		$user = elgg_get_logged_in_user_entity();
+	}
+	
+	if (empty($user)) {
```
Comment (jeabakker) [inline mod/groups/start.php]: same as above

File: mod/groups/start.php [hunk 4]
```diff
@@ -357,6 +355,82 @@ function groups_activity_owner_block_menu($hook, $type, $return, $params) {
...
+	if (!$group->isMember($user) || ($group->owner_guid === $user->guid)) {
+		// a member can leave a group if he/she doesn't own it
+		return false;
+	}
+	
+	return \ElggMenuItem::factory([
+		'name' => 'groups:leave',
+		'icon' => 'sign-out',
+		'text' => elgg_echo('groups:leave'),
+		'href' => "action/groups/leave?group_guid={$group->guid}",
```
Comment (jeabakker) [inline mod/groups/start.php]: add user_guid to the url as apparently this function is allowed for another user than current user

File: mod/groups/start.php [hunk 5]
```diff
@@ -357,6 +355,82 @@ function groups_activity_owner_block_menu($hook, $type, $return, $params) {
...
+ * @param \ElggUser  $user  User to check leave action for
+ *
+ * @return \ElggMenuItem|false
+ */
+function groups_get_group_join_menu_item(\ElggGroup $group, $user = null) {
+	if (!$group instanceof \ElggGroup) {
+		return false;
+	}
+	
+	if (is_null($user)) {
```
Comment (jeabakker) [inline mod/groups/start.php]: same user check as before

File: mod/groups/start.php [hunk 6]
```diff
@@ -357,6 +355,82 @@ function groups_activity_owner_block_menu($hook, $type, $return, $params) {
...
+		$user = elgg_get_logged_in_user_entity();
+	}
+	
+	if (empty($user)) {
```
Comment (jeabakker) [inline mod/groups/start.php]: ....

File: mod/groups/start.php [hunk 7]
```diff
@@ -357,6 +355,82 @@ function groups_activity_owner_block_menu($hook, $type, $return, $params) {
...
+		// admins can always join
+		// non-admins can join if membership is public
+		$menu_name = 'groups:join';
+	}
+	
+	return \ElggMenuItem::factory([
+		'name' => $menu_name,
+		'icon' => 'sign-in',
+		'text' => elgg_echo($menu_name),
+		'href' => "action/groups/join?group_guid={$group->guid}",
```
Comment (jeabakker) [inline mod/groups/start.php]: add user_guid

File: mod/groups/start.php [hunk 8]
```diff
@@ -373,32 +447,45 @@ function groups_entity_menu_setup($hook, $type, $return, $params) {
...
+	if (!empty($group_join)) {
+		$return[] = $group_join;
+	}
+	
+	$group_leave = groups_get_group_leave_menu_item($entity, $user);
+	if (!empty($group_leave)) {
+		$return[] = $group_leave;
+	}
+		
+	if (elgg_is_admin_logged_in()) {
```
Comment (jeabakker) [inline mod/groups/start.php]: you already have $user so why not check $user->isAdmin()

File: mod/groups/views/default/groups/profile/fields.php [hunk 9]
```diff
@@ -3,32 +3,43 @@
...
+	if ($key == 'name') {
+		continue;
+	}
 
-		$options = ['value' => $group->$key];
-		if ($valtype == 'tags') {
-			$options['tag_names'] = $key;
-		}
+	$value = $group->$key;
+	if (is_null($value)) {
```
Comment (jeabakker) [inline mod/groups/views/default/groups/profile/fields.php]: do we also want to skip `''`?

File: mod/members/start.php [hunk 10]
```diff
@@ -17,6 +17,16 @@ function members_init() {
 		'href' => 'members',
 	]);
 
+	if (elgg_is_admin_logged_in()) {
+		elgg_register_menu_item('title', [
+			'name' => 'add_user',
+			'text' => elgg_echo('admin:users:add'),
+			'href' => 'admin/users/add',
+			'context' => 'members',
+			'class' => 'elgg-button elgg-button-action',
```
Comment (jeabakker) [inline mod/members/start.php]: `link_class`

**Review by jeabakker — DISMISSED** [review]

File: mod/groups/start.php (hunk 4 above)
Comment (hypeJunction) [reply mod/groups/start.php]: elgg_http_add_url_query_elements

File: mod/groups/views/default/groups/profile/fields.php (hunk 1 above)
Comment (jdalsem) [reply mod/groups/views/default/groups/profile/fields.php]: should be done as part of a fix for #10412.. this PR makes them at least inline with eachother

File: mod/groups/views/default/groups/profile/fields.php (hunk 9 above)
Comment (jdalsem) [reply mod/groups/views/default/groups/profile/fields.php]: maybe... but leaving it as is for now... empty strings are not saved when editing the profile

**Pull request merged**